Blog > Application: 블로그 수정 repository port, use case, service

merge-simpson opened the pull request:
## Issues

- Resolves #54 

## Description

- 블로그 기본 정보 수정
- 다음 작업의 하위 작업  
    #9 > #43 

<br />

## Review Points

- 테스트 케이스 위주로 살펴 주시면 좋겠습니다. 감사합니다.

<br />

## How Has This Been Tested?

- ChatGPT > 새 프로젝트 생성 > 테스트 코드 **<ins>예시 기반</ins> 지침 추가**, 추가적인 제안 허용  
    <img width="850" height="284" alt="ChatGPT 프로젝트에서 테스트코드 요청 화면 캡처" src="https://github.com/user-attachments/assets/381da43f-ebe9-4888-b860-ac47b0090f0c" />
- **Test cases**  
  <img width="430" height="97" alt="인텔리제이 UI에서 테스트 케이스 캡처" src="https://github.com/user-attachments/assets/c7a180e5-f0e6-4dc2-9170-6aba520b5643" />

merge-simpson committed:
feat(blog): add `deleteById` in repository port

merge-simpson committed:
feat(blog): implement `deleteById` in service

merge-simpson committed:
test(blog): add test for successful deletion

verify that deleting an existing blog decrements list size and invokes deleteById once

merge-simpson committed:
test(blog): add test for exception on deleting non-existent blog

mock deleteById to throw BLOG_NOT_FOUND exception and verify deleteById not called for wrong id

merge-simpson committed:
feat(blog): add `existsById` method to `BlogCommandRepositoryPort`

enable checking blog existence by id before operations like delete, update or create

merge-simpson committed:
feat(blog): implement update operation in BlogCommandService

- check blog existence with existsById and throw BLOG_NOT_FOUND.exception if absent
- build updated Blog entity using builder and save via commandRepository.save

merge-simpson committed:
feat(blog): add findById method to BlogCommandRepositoryPort

enable fetching blog by id returning Optional<Blog>

merge-simpson committed:
fix(blog): require non-null blogId and switch update to patch-style using findById

- add Objects.requireNonNull(blogId) validation
- default null name/url to empty string
- replace existsById check with findById and prepareUpdate().update() pattern

merge-simpson committed:
feat(blog): add error code for blank blog name

add BLOG_NAME_CANNOT_BE_BLANK to BlogErrorCode enum for validating non-empty blog name

merge-simpson committed:
fix(blog): enforce non-blank name in update

replace null-or-blank name fallback with BLOG_NAME_CANNOT_BE_BLANK exception instead of default empty string

merge-simpson committed:
test(blog): add update tests for blog service

cover successful update, name null validation error, and null url no-change

wowddok99 reviewed (approved):
올려주신 코드 확인하였습니다. 깔끔한 작업 감사드립니다!
추후에 저 포함 다른 분들도 해당 코드 참고해서 테스트 코드 짜보면 될듯합니다!!
고생하셨습니다😄

sweetykr7 commented:
혹시 지침을 어떻게 넣으셨는지 알 수 있을까요?

merge-simpson commented:
@sweetykr7: (_origin: https://github.com/nettee-space/nettee-blolet-backend/pull/62#issuecomment-3111888276_)

> 혹시 지침을 어떻게 넣으셨는지 알 수 있을까요?

## <!-- thin -->

@merge-simpson:

프롬프팅이 전문영역이 아니어서 미숙한 부분이 있겠습니다만, 다음처럼 넣었습니다.
개선이 필요한 부분을 발견하시면 언급 부탁드립니다:

````markdown
Service 로직에 대한 테스트

- Production code is written in Java.
- Test code is written in Kotlin.
- Command 관련 테스트는 기존 양식을 참고:

```kotlin
import io.kotest.assertions.throwables.*
import io.kotest.core.spec.style.FreeSpec
import io.kotest.matchers.*
import io.kotest.matchers.equals.*
import io.mockk.*
import nettee.blolet.blog.application.port.BlogCommandRepositoryPort
import nettee.blolet.blog.domain.Blog
import nettee.blolet.blog.exception.BlogErrorCode.BLOG_MAXIMUM_EXCEEDED
import nettee.blolet.blog.exception.BlogErrorCode.BLOG_NOT_FOUND
import nettee.common.CustomException
import java.time.Instant

class BlogCommandServiceTest : FreeSpec({
    val commandPort = mockk<BlogCommandRepositoryPort>()
    val commandService = BlogCommandService(commandPort)

    beforeTest {
        clearMocks(commandPort, answers = true, recordedCalls = true)
    }

    "[UPDATE] 블로그 수정 시" - {
        val targetId = "1"
        val newName = "new blog name"
        val newUrl = "https://user1.blolet.com"

        // mock 데이터 초기화
        val now = Instant.now()
        val original = Blog.builder()
            .id(targetId)
            .userId("USER-1")
            .name("A Blog")
            .url("https://old.blolet.com")
            .createdAt(now)
            .updatedAt(now)
            .build()
        val captured = mutableListOf<Blog>()

        beforeTest {
            clearMocks(commandPort, answers = true, recordedCalls = true)
            // 조회 시 원본 반환
            every { commandPort.findById(targetId) } returns Optional.of(original)
            // 저장 시 capture 후, updatedAt만 갱신된 새 객체 반환
            every { commandPort.save(capture(captured)) } answers {
                val input: Blog = firstArg()
                Blog.builder()
                    .id(input.id!!)
                    .userId(input.userId)
                    .name(input.name)
                    .url(input.url)
                    .createdAt(input.createdAt)
                    .updatedAt(Instant.now())
                    .build()
            }
        }

        "✅ 존재하는 아이템을 수정할 수 있다." {
            // action
            val updated: Blog = commandService.update(targetId, newName, newUrl)

            // assert
            updated.id shouldBe targetId
            updated.userId shouldBe original.userId
            updated.name shouldBe newName
            updated.url shouldBe newUrl

            // 호출 순서 확인
            verifySequence {
                commandPort.findById(targetId)
                commandPort.save(any())
            }
        }

        "🚧 name은 null이어선 안 된다." {
            // action & assert
            val exception = shouldThrow<CustomException> {
                commandService.update(targetId, null, newUrl)
            }
            exception.errorCode shouldBe BLOG_NAME_CANNOT_BE_BLANK
            verify(inverse = true) { commandPort.save(any()) }
            // `findById` 메서드는 호출되든 아니든 영향이 없음. (테스트 대상 X)
            //  ❌ verify(inverse = true) { commandPort.findById(targetId) }
        }

        "✅ 입력 URL이 null이면 URL은 수정되지 않는다." {
            // action
            val updated: Blog = commandService.update(targetId, newName, null)

            // assert
            updated.name shouldBe newName
            updated.url shouldBe original.url // URL 변경 없음

            verifySequence {
                commandPort.findById(targetId)
                commandPort.save(any())
            }
        }
    }
})
```

응답:

FreeSpec과 kotest를 활용하여 위 양식에 유사한 테스트 코드를 제안한다.
단, 더 나은 양식이 있다면 제안해도 된다.
````
